### PR TITLE
feat(gyp): update gyp to v0.16.1

### PR DESCRIPTION
**Pull request by rzhao271**
##### Checklist

- [x] `npm install && npm test` passes
- [x] tests are included 
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/googleapis/release-please#how-should-i-write-my-commits)

### CODE REVIEW

**Comment by rzhao271** [comment]
@cclauss one issue I have with this PR is that the Python portion of node-gyp actually requires a package now. I'm wondering whether we could change the install script to set up the Python package for the user, or will users have to install the required Python packages themselves?

I'll also have to take a look at the VS tests tomorrow.

**Comment by cclauss** [comment]
> the Python portion of node-gyp actually requires a package now

We vendor in this code into the node-gyp repo so that users do need to pip install it.  We post [gyp-next to PyPI](https://pypi.org/project/gyp-next/) but [statistics show](https://pypistats.org/packages/gyp-next) that this is not how most users access it.  I do not see how anything in this release changes that process.

**Comment by rzhao271** [comment]
Do the 2016 and 2019 Visual Studio tests still make sense?
When I expand the Environment Information tab, the only IDE listed is VS 2022.

The failing tests also emit the following output:
```
gyp verb find VS msvs_version not set from command line or npm config
+ gyp verb find VS VCINSTALLDIR not set, not running in VS Command Prompt
+ gyp verb find VS checking VS2022 (17.7.34031.279) found at:
+ gyp verb find VS "C:\Program Files\Microsoft Visual Studio\2022\Enterprise"
+ gyp verb find VS - found "Visual Studio C++ core features"
+ gyp verb find VS - found VC++ toolset: v143
+ gyp verb find VS - found Windows SDK: 10.0.22621.0
+ gyp info find VS using VS2022 (17.7.34031.279) found at:
+ gyp info find VS "C:\Program Files\Microsoft Visual Studio\2022\Enterprise"
```

**Comment by cclauss** [comment]
You are saying that this environment variable trick no longer works?

https://github.com/nodejs/node-gyp/blob/b3d41aeb737ddd54cc292f363abc561dcc0a614e/.github/workflows/visual-studio.yml#L28

**Comment by cclauss** [comment]
% `npm config list -l`

npm recently got rid of lots of config options including `msvs-version`

**Comment by rzhao271** [comment]
`VCINSTALLDIR` works locally
`GYP_MSVS_VERSION` does not

**Comment by DeeDeeG** [comment]
> one issue I have with this PR is that the Python portion of node-gyp actually requires a package now.

This refers to the `packaging` (https://pypi.org/project/packaging/) dependency, right?

> We vendor in this code into the node-gyp repo so that users do need to pip install it.

This refers to `gyp-next` itself, right?

...

Just to clarify, there is a new requirement (a Python package called `packaging`) that is needed to run this, and which  users won't have on their system by default?

I don't think `gyp-next` has vendored in the `packaging` dependency... So, this updated `gyp-next` wouldn't run out-of-the-box on a system with just Python, right? **Is it an option to vendor in a copy of `packaging` into `gyp-next` repo?** That would solve the problem, if I've understood the problem correctly in the first place.

**Comment by cclauss** [comment]
Is `packaging` a runtime requirement or just a test-time requirement?

File: gyp/pylib/gyp/input.py [hunk 1]
```diff
@@ -16,9 +16,9 @@
 import sys
 import threading
 import traceback
-from distutils.version import StrictVersion
 from gyp.common import GypError
 from gyp.common import OrderedSet
+from packaging.version import Version
```
Comment (DeeDeeG) [inline gyp/pylib/gyp/input.py:21]: Yes, the `packaging` dependency is run-time required here, not just in testing.

<details><summary>Error log (click to expand):</summary>

```
# ~/node-gyp/bin/node-gyp.js rebuild          
gyp info it worked if it ends with ok
gyp info using node-gyp@9.4.0
gyp info using node@16.20.2 | linux | x64
undefined
gyp info find Python using Python version 3.10.12 found at "/usr/bin/python3"
gyp info spawn /usr/bin/python3
gyp info spawn args [
gyp info spawn args '/root/node-gyp/gyp/gyp_main.py',
gyp info spawn args 'binding.gyp',
gyp info spawn args '-f',
gyp info spawn args 'make',
gyp info spawn args '-I',
gyp info spawn args '/root/require_superstring/node_modules/superstring/build/config.gypi',
gyp info spawn args '-I',
gyp info spawn args '/root/node-gyp/addon.gypi',
gyp info spawn args '-I',
gyp info spawn args '/root/.cache/node-gyp/16.20.2/include/node/common.gypi',
gyp info spawn args '-Dlibrary=shared_library',
gyp info spawn args '-Dvisibility=default',
gyp info spawn args '-Dnode_root_dir=/root/.cache/node-gyp/16.20.2',
gyp info spawn args '-Dnode_gyp_dir=/root/node-gyp',
gyp info spawn args '-Dnode_lib_file=/root/.cache/node-gyp/16.20.2/<(target_arch)/node.lib',
gyp info spawn args '-Dmodule_root_dir=/root/require_superstring/node_modules/superstring',
gyp info spawn args '-Dnode_engine=v8',
gyp info spawn args '--depth=.',
gyp info spawn args '--no-parallel',
gyp info spawn args '--generator-output',
gyp info spawn args 'build',
gyp info spawn args '-Goutput_dir=.'
gyp info spawn args ]
Traceback (most recent call last):
  File "/root/node-gyp/gyp/gyp_main.py", line 42, in <module>
    import gyp  # noqa: E402
  File "/root/node-gyp/gyp/pylib/gyp/__init__.py", line 9, in <module>
    import gyp.input
  File "/root/node-gyp/gyp/pylib/gyp/input.py", line 21, in <module>
    from packaging.version import Version
ModuleNotFoundError: No module named 'packaging'
gyp ERR! configure error 
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack at ChildProcess.onCpExit (/root/node-gyp/lib/configure.js:305:16)
gyp ERR! stack at ChildProcess.emit (node:events:513:28)
gyp ERR! stack at Process.ChildProcess._handle.onexit (node:internal/child_process:293:12)
gyp ERR! System Linux 5.15.49-linuxkit
gyp ERR! command "/usr/local/bin/node" "/root/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /root/require_superstring/node_modules/superstring
gyp ERR! node -v v16.20.2
gyp ERR! node-gyp -v v9.4.0
gyp ERR! not ok 
```

</details>

**Comment by rzhao271** [comment]
@cclauss `packaging` is now a runtime requirement for node-gyp, but I have no experience with vendoring. Would it make sense for the repository to include a copy of the packaging package's source?

**Comment by DeeDeeG** [comment]
I think this PR is obsolete and can be superseded by a bump directly to gyp-next 0.16.0.

If someone who actually worked hard on these changes in gyp-next wants to open a PR bumping straight to gyp-next 0.16.0, I'll let you do that. Since it will show the commits here with the author as whoever performs the gyp-next bump.

Otherwise, I would want to open such a PR bumping straight to gyp-next 0.16.0 myself soon.

Thanks for the hard work in this, folks! It will be both exciting and a relief to have node-gyp working on Python 3.12.

**Comment by rzhao271** [comment]
@DeeDeeG go ahead with the bump to v0.16.0
There is a script you can run in this repository to help create the commit
After that, you can copy over the CI changes from this commit.

**Comment by cclauss** [comment]
Where are we at on this?  Let's try to get [Py3.12 compatibility](https://github.com/nodejs/node-gyp/issues/2869#issuecomment-1775625254) into @lukekarrys imminent releases.
https://github.com/nodejs/node-gyp/issues/2860#issuecomment-1783189734

**Comment by lukekarrys** [comment]
There will be an automated release PR for `v10.0.0` where we can discuss waiting for any open PRs to land first. I won't release v10 without waiting a few days to make sure we get everything in.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -40,7 +40,7 @@ jobs:
       - name: Install Dependencies
         run: |
           npm install --no-progress
-          pip install pytest
+          python -m pip install pytest packaging
```
Comment (cclauss) [inline .github/workflows/tests.yml:43]: We can revert this change because of
* nodejs/gyp-next#214

**Comment by rzhao271** [comment]
Closing in favour of https://github.com/nodejs/node-gyp/pull/2923 because it has a better description
